### PR TITLE
Features/encoder service

### PR DESCRIPTION
**Pull request by PabloGonzalezSegarra**
# Finished encoder service

The service works fine and has been fully documented using doxygen. As you can see below:

captura

## Testing:

The class has been tested using the waveforms generating two signals 90 degrees out of phase with each other, it was verified that when exchanging the offset it changed direction as it should. As can be seen in the following image.

image

In the nucleo, a led blinked every time the CNT changed and another led was on or off depending on the direction and the correct operation was also verified using the debugger and the live expressions through the code shown below:

captura

### CODE REVIEW

File: Inc/HALAL/Services/Encoder/Encoder.hpp [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+ * The first pin of the structure
+ * @var DoublePin::pin2
+ * The second pin of the structure
+ *
+ * */
+struct DoublePin {
+	Pin* pin1;
+	Pin* pin2;
+
+	bool operator<(const DoublePin& other) const {
```
Comment (alejandrogonzalvo) [inline Inc/HALAL/Services/Encoder/Encoder.hpp:26]: Possible refactor with ternary operator, is equivalent.
```suggestion
	bool operator<(const DoublePin& other) const {
	    return pin1 < other.pin1 ? true : pin1 == other.pin1 ? pin2 < other.pin2 : false;
```

File: Inc/HALAL/Services/Encoder/Encoder.hpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+ *      Author: Pablo
+ */
+
+#pragma once
+#ifdef HAL_TIM_MODULE_ENABLED
+#include "ST-LIB.hpp"
+
+/**
+ * @struct DoublePin
+ * @brief Estruct that contains a pair of pins
```
Comment (alejandrogonzalvo) [inline Inc/HALAL/Services/Encoder/Encoder.hpp:14]:
```suggestion
 * @brief Struct that contains a pair of pins
```

File: Inc/HALAL/Services/Encoder/Encoder.hpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+};
+
+struct EncoderPinTimer{
+	Pin pin1;
+	Pin pin2;
+	TIM_HandleTypeDef* timer;
+};
+
+/**
+ * @brief Encoder service class for abstracting the use of the encoder with the HAL library.
```
Comment (alejandrogonzalvo) [inline Inc/HALAL/Services/Encoder/Encoder.hpp:64]: Short sentences are more legible.
```suggestion
 * @brief Encoder service class. Abstracts the use of the encoder with the HAL library.
```

File: Inc/HALAL/Services/Encoder/Encoder.hpp [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+ *
+ */
+class Encoder {
+public:
+	static forward_list<uint8_t> ID_manager;
+	static map<DoublePin, TIM_HandleTypeDef*> pin_timer_map;
+	static map<uint8_t, DoublePin> service_IDs;
+
+	/**
+	 * @brief This method register a new encoder
```
Comment (alejandrogonzalvo) [inline Inc/HALAL/Services/Encoder/Encoder.hpp:74]:
```suggestion
	 * @brief This method registers a new encoder
```

**Review by alejandrogonzalvo — APPROVED** [review]
Aprobada pero con feedback de faltas de ortografia y un posible refactor, revisar

File: Inc/HALAL/Services/Encoder/Encoder.hpp (hunk 1 above)
Comment (StefanCostea) [reply Inc/HALAL/Services/Encoder/Encoder.hpp:26]: That's horrible looking

Comment (PabloGonzalezSegarra) [reply Inc/HALAL/Services/Encoder/Encoder.hpp:26]: I dont like it either.

**Comment by PabloGonzalezSegarra** [comment]
No aceptéis el pull request todavía, he encontrado algunos fallos.

**Review by StefanCostea — CHANGES_REQUESTED** [review]

**Comment by PabloGonzalezSegarra** [comment]
Ya esta todo arreglado y funcional.

**Comment by StefanCostea** [comment]
Are you still using DoublePin struct or not?

**Review by StefanCostea — APPROVED** [review]

**Pull request merged**